### PR TITLE
chore: Upgrade ESLint to v9 and overhaul configuration

### PR DESCRIPTION
**Pull request by RubenSibon**
# Describe the pull request

Upgrade ESLint to version 9 and overhaul the configuration for ESLint and it's plugin ecosystem.

## What

- Upgrades ESLint to version 9.x;
- Upgrades various packages related to ESLint;
- Introduces the [flat config file for ESLint](https://eslint.org/blog/2022/08/new-config-system-part-2/);
- Overhauls configuration files for various ESLint plugins and TypeScript;
- Addresses a number of lint issues that came up.

## Why

ESLint v8.x reached end-of-life on 2024-10-05 and is no longer maintained.

## How

By following the [migration guide](https://eslint.org/docs/latest/use/migrate-to-9.0.0).

## Checklist

Before submitting your pull request, please ensure you have done the following. Check each checkmark if you have done so or if it wasn't necessary:

- [x] Start the PR title with a Conventional Commit prefix, [as explained here](https://github.com/Amsterdam/design-system/blob/main/documentation/publishing.md?plain=1#L11).

## Additional notes

- Important: please check if linting is happening properly and that there are no false negatives or false positives.
- ESLint migration documentation:
  - [Migrate to v9.x](https://eslint.org/docs/latest/use/migrate-to-9.0.0)
  - [Configuration Migration Guide](https://eslint.org/docs/latest/use/configure/migration-guide)

### CODE REVIEW

File: CONTRIBUTING.md [hunk 1]
```diff
@@ -140,6 +140,8 @@ You can use any editor you like, but if you use [Visual Studio Code](https://cod
 
 To enable correct validation and to fix lint/style errors on save, add this to your VSCode `settings.json`:
 
+<!-- eslint-disable -->
```
Comment (alimpens) [inline CONTRIBUTING.md:143]: Maybe we can just add colons around the JSON block, to make it valid JSON?

File: .eslintrc.ts.json [hunk 2]
```diff
@@ -1,6 +1,5 @@
 {
   "rules": {
-    "@typescript-eslint/consistent-type-definitions": ["error", "type"],
```
Comment (alimpens) [inline .eslintrc.ts.json:3]: We do want this rule, right?

File: proprietary/react-icons/README.md [hunk 3]
```diff
@@ -19,13 +19,11 @@ npm install @amsterdam/design-system-react-icons
 
 Import the component for the icon you need, as well as the generic Icon component, and use them in your JSX.
```
Comment (alimpens) [inline proprietary/react-icons/README.md:21]: Maybe we can ignore Prettier here, to keep the old formatting? Specifically the single quotes and not using semicolons.

File: storybook/src/components/SearchField/SearchField.stories.tsx [hunk 4]
```diff
@@ -75,7 +75,6 @@ export const Controlled: any = {
           const value = formData.get('search-box')
           // search actions should not be triggered without a value
           if (value) {
-            // eslint-disable-next-line no-alert
```
Comment (alimpens) [inline storybook/src/components/SearchField/SearchField.stories.tsx:78]: I think this should trigger an ESLint warning? We don't want to use `alert` in our code, I think?

File: CONTRIBUTING.md (hunk 1 above)
Comment (RubenSibon) [reply CONTRIBUTING.md:142]: Done

File: eslint.config.mjs [hunk 5]
```diff
@@ -0,0 +1,109 @@
...
+    ...json.configs.recommended,
+  },
+
+  // Markdown
+  {
+    plugins: { markdown },
+  },
+  {
+    files: ['**/*.md'],
+    ignores: ['dist/**/*.md', 'CHANGELOG.md'],
```
[the remark this answers is not in the document]
Comment (RubenSibon) [reply eslint.config.mjs]: Removed.

File: proprietary/react-icons/README.md (hunk 3 above)
Comment (RubenSibon) [reply proprietary/react-icons/README.md:21]: Done

File: eslint.config.mjs [hunk 6]
```diff
@@ -0,0 +1,109 @@
...
+        ecmaFeatures: { jsx: true },
+      },
+    },
+    settings: {
+      'import/resolver': {
+        node: {
+          extensions: ['js', 'mjs', 'jsx', 'ts', 'tsx'],
+        },
+      },
+      react: { version: '18' },
```
[the remark this answers is not in the document]
Comment (RubenSibon) [reply eslint.config.mjs]: Done, before it complained that it couldn't find React. But I've not seen that issue anymore.

File: eslint.config.mjs [hunk 7]
```diff
@@ -0,0 +1,109 @@
...
+    processor: mdx.createRemarkProcessor({
+      lintCodeBlocks: true,
+    }),
+  },
+  ...compat.extends('plugin:mdx/overrides').map((config) => ({
+    ...config,
+    ...mdx.flatCodeBlocks,
+    rules: {
+      ...mdx.flatCodeBlocks.rules,
+      'react/jsx-no-undef': 'off',
```
[the remark this answers is not in the document]
Comment (RubenSibon) [reply eslint.config.mjs:233]: Removed.

File: .eslintrc.ts.json (hunk 2 above)
Comment (RubenSibon) [reply .eslintrc.ts.json:3]: Added again.

File: storybook/src/components/SearchField/SearchField.stories.tsx (hunk 4 above)
Comment (RubenSibon) [reply storybook/src/components/SearchField/SearchField.stories.tsx]: Restored.

File: eslint.config.mjs [hunk 8]
```diff
@@ -0,0 +1,235 @@
...
+  baseDirectory: __dirname,
+  recommendedConfig: eslint.configs.recommended,
+  allConfig: eslint.configs.all,
+})
+
+export default tseslint.config(
+  // Global
+  {
+    ignores: [
+      '**/node_modules/',
```
Comment (alimpens) [inline eslint.config.mjs:29]: I don't think this is necessary, this is part of [the default patterns](https://eslint.org/docs/latest/use/configure/ignore#ignoring-files)

File: eslint.config.mjs [hunk 9]
```diff
@@ -0,0 +1,235 @@
...
+      'no-throw-literal': 'error',
+      'no-undef': 'error',
+      'no-unexpected-multiline': 'error',
+      'no-unmodified-loop-condition': 'error',
+      'no-unreachable': 'error',
+      'no-unsafe-finally': 'error',
+      'no-unsafe-negation': 'error',
+      'no-unused-expressions': 'error',
+      'no-unused-labels': 'error',
+      '@typescript-eslint/no-unused-vars': 'error',
```
Comment (alimpens) [inline eslint.config.mjs:165]: Shouldn't this be at the top?

File: packages/react/documentation/coding-conventions.md [hunk 10]
```diff
@@ -9,11 +9,10 @@ Barrel files allow a consumer to do this:
 
 Instead of a separate import for each component, while even reaching into the `dist` directory:
 
-<!-- prettier-ignore -->
```
Comment (alimpens) [inline packages/react/documentation/coding-conventions.md:12]: We want to keep this, I think? So we can get rid of the semicolons?

File: storybook/src/docs/developer-guide/getting-started.docs.mdx [hunk 11]
```diff
@@ -16,7 +16,7 @@ npm install @amsterdam/design-system-assets @amsterdam/design-system-css @amster
 
 Import the components and stylesheets you need, for example:
 
-```javascript
+```tsx
```
Comment (alimpens) [inline storybook/src/docs/developer-guide/getting-started.docs.mdx:19]: We can ignore prettier here and get rid of the semicolons

**Review by alimpens — APPROVED** [review]

**Review by VincentSmedinga — APPROVED** [review]
Good job! 🎉

**Pull request merged**